### ISSUE TITLE
ExportItems and UploadItems

### ISSUE THREAD

**Issue by LukeMarlin** [OfficeDev/ews-managed-api#46]
Hi guys,

Do you intend to implement these two methods in EWS managed API, or will we have to stick with web calls ?

**Comment by dclaux** [comment]
No plan to do this at this time, but if you want to give it a shot you are more than welcome.

**Comment by XavierDa** [comment]
Hello, i'm going to implement it. (i'm luke co-worker)

Just to be sure i need to : 
- Implement Responses
- Implement request 
- add new functions to service 

Do this sound right ?

**Comment by LukeMarlin** [comment]
Any insight from you @dclaux @vboctor  ?

**Comment by dclaux** [comment]
@XavierDa Yes, that sounds about right. Looking forward to looking at your changes.

**Comment by XavierDa** [comment]
Hello, i finished implementing it, but i have a small problem.

It seems that exportItems change ItemIds (can't find any of the itemids i sent in the result), Is there any way to find a link between them ?

**Comment by dclaux** [comment]
One caveat - The typical use case for UploadItems is read data from somewhat large files (as opposed to in-memory buffers) or from a database field. So the EWS Managed API's implementation of UploadItems should stream the data right from its source and serialize it by chunks, so that there is no need to load a full copy of all the data in memory. You can take a look at how file attachments are handled to see an example of the how the API does this.

**Comment by dclaux** [comment]
ExportItems shouldn't change any Id. It's just a dump of what's in a mailbox.

**Comment by dclaux** [comment]
So are you saying that the Ids you get in ExportItems' response aren't the same as the ones you passed in the request? Sounds really weird to me. That said the items in the response should be in the same order as those in the request, so you can match based on ordering.

**Comment by XavierDa** [comment]
Ok found my error. I had a misplaced new ItemId.

For uploadItems... i will check how it's done for file attachement and try to get a similar solution.
Right now my solution for exportItem returns a byte[] for data, maybe it should use a Stream too ? (so it's easier to implement export => upload solutions ?)

**Comment by dclaux** [comment]
Yes, ExportItems should also use streams.

**Comment by dclaux** [comment]
Thinking more about this - ExportItems needs to be passed a callback method that lets the developer specify the stream for each item in the response. Either that, or introduce an "ExportManager" class that encapsulates the entire logic and exposes an event that lets developers specify the destination stream for each item. I can think of two main ways a developer might want to store what ExportItems produces: 1) one file/database row/field per item, or 2) a single file that stores all items. I think both scenarios should be easily doable.

**Comment by Biztactix-Ryan** [comment]
How far did you get with this? @XavierDa

**Comment by XavierDa** [comment]
We have a working version here, but we never manage to make a PR due to some slight hiccups on our side.

if you want this you can get it from this pull request https://github.com/OfficeDev/ews-managed-api/pull/109

**Comment by jrouzies-fr** [comment]
@XavierDa , are we sure it is working?

I can get the data without issue, but the UploadItem doesn't seem to do anything on the target mailbox / service

![image](https://user-images.githubusercontent.com/57990274/73443945-4dd5cc80-4360-11ea-813c-b8e89757c648.png)

Source mailbox Inbox:

![image](https://user-images.githubusercontent.com/57990274/73444010-6cd45e80-4360-11ea-8dde-7ac2a1a6d5f0.png)

Target mailbox Inbox:

![image](https://user-images.githubusercontent.com/57990274/73444050-81b0f200-4360-11ea-89f4-a3b465e18474.png)

**Comment by XavierDa** [comment]
Did you check your uploaditem returns ? it may succeed globally but each item may fail.

**Comment by jrouzies-fr** [comment]
**(fix for my issue in Update2 below)**

> Did you check your uploaditem returns ? it may succeed globally but each item may fail.

Indeed my bad, there is an error in the response, indicating malformed ID:

![image](https://user-images.githubusercontent.com/57990274/73444593-8de97f00-4361-11ea-8452-9997b280209c.png)

Is it possibles IDs from Exchange Online are incompatible with OnPremise ones?

_________________________________________________________

**Update**: tried from OnPrem to OnPrem, same issue. Realized ItemToUpload ID was not filled, don't know if it was on purpose. But even filling it still shows the issue:

![image](https://user-images.githubusercontent.com/57990274/73445640-d1dd8380-4363-11ea-94c0-acdb0071f48d.png)

_________________________________________________________

**Update 2**: I made further tests, and I created a custom folder, and search for that specific folder ID before using it in the itemToUpload. It worked.

It seems what is causing the issue is the "WellKnownFolderName.Inbox".

![image](https://user-images.githubusercontent.com/57990274/73450964-289c8a80-436f-11ea-940a-30c14a6ff2a4.png)

**To conclude**: make sure you get the ID by binding the folder in the destination service, not just by specifying a WellKnownFolder in the item to upload.

**Comment by aldrinbucagahsay** [comment]
Hi, I am also using the ExportItems and UploadItems to backup and restore items in outlook. Both are working, however when I use the CreateAction Update to update an existing item, the "pin/unpin" and "mark as read/unread" property of the item in UI is not updated. Though these fields are updated when I use CreateNew as long as the item does not exist in Inbox. Does anyone encountered this issue or anyone who knows how to fix it? Thanks!
